### PR TITLE
data processing document: add tf pytorch dataframe examples

### PR DESCRIPTION
**Pull request by yangw1234**

### CODE REVIEW

File: docs/docs/Orca/data-parallel-processing.md [hunk 1]
```diff
@@ -3,20 +3,104 @@
 ### **1. TensorFlow Dataset and PyTorch DataLoader**
 
 Orca will seamlessly parallelize the standard `tf.data.Dataset` or `torch.utils.data.DataLoader` pipelines across a large cluster in a data-parallel fashion, which can be directly used for distributed deep learning training, as shown below:
-<TODO: shown a simple example>
+
+TensorFlow:
+```python
+import tensorflow as tf
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:9]: import tensorflow_datasets as tfds

File: docs/docs/Orca/data-parallel-processing.md [hunk 2]
```diff
@@ -3,20 +3,104 @@
...
+
+def preprocess(data):
+    data['image'] = tf.cast(data["image"], tf.float32) / 255.
+    return data['image'], data['label']
+
+train_dataset = tfds.load(name="mnist", split="train", data_dir=dataset_dir)
+train_dataset = train_dataset.map(preprocess)
+
+est = Estimator.from_keras(keras_model=model)
+est.fit(data=train_dataset,
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:20]: Just `est.fit(data=train_dataset)`

File: docs/docs/Orca/data-parallel-processing.md [hunk 3]
```diff
@@ -3,20 +3,104 @@
...
+        batch_size=320,
+        epochs=1)
+```
+
+Pytorch:
+```python
+import torch
+import torch.nn as nn
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:28]: not needed?

File: docs/docs/Orca/data-parallel-processing.md [hunk 4]
```diff
@@ -3,20 +3,104 @@
...
+import torch.nn.functional as F
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:29]: not needed?

File: docs/docs/Orca/data-parallel-processing.md [hunk 5]
```diff
@@ -3,20 +3,104 @@
 ### **1. TensorFlow Dataset and PyTorch DataLoader**
 
 Orca will seamlessly parallelize the standard `tf.data.Dataset` or `torch.utils.data.DataLoader` pipelines across a large cluster in a data-parallel fashion, which can be directly used for distributed deep learning training, as shown below:
-<TODO: shown a simple example>
+
+TensorFlow:
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:7]: TensforFlow Dataset

File: docs/docs/Orca/data-parallel-processing.md [hunk 6]
```diff
@@ -3,20 +3,104 @@
...
+train_dataset = tfds.load(name="mnist", split="train", data_dir=dataset_dir)
+train_dataset = train_dataset.map(preprocess)
+
+est = Estimator.from_keras(keras_model=model)
+est.fit(data=train_dataset,
+        batch_size=320,
+        epochs=1)
+```
+
+Pytorch:
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:25]: PyTorch DataLoader

File: docs/docs/Orca/data-parallel-processing.md [hunk 7]
```diff
@@ -3,20 +3,104 @@
...
+zoo_estimator.fit(data=train_loader, epochs=1)
+```
+
+
 
 Under the hood, Orca will automatically replicate the _TensorFlow Dataset_ or _PyTorch DataLoader_ pipeline on each node in the cluster, shard the input data, and execute the data pipelines using Apache Spark and/or Ray distributedly. 
 
-**Note:** Known limitations include: <TODO: describe known limitations>
+**Note:** Known limitations include:
+1. TensorFlow Dataset pipeline that contains transformations defined in native python function, such as `tf.py_func`, `tf.py_function
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:50]: `tf.py_function`

File: docs/docs/Orca/data-parallel-processing.md [hunk 8]
```diff
@@ -3,20 +3,104 @@
...
+and `tf.numpy_function` are currently not supported.
+2. TensorFlow Dataset pipeline created from generators, such as `Dataset.from_generators` are currently not supported.
+3. TensorFlow Dataset pipeline that are file based, such as `TFRecordDataset` and `TextLineDataset`. The files to be read should be either
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:53]: Combine 3 & 4:

For TensorFlow Dataset and Pytorch DataLoader pipelines that read from files (including `tf.data.TFRecordDataset` and `tf.data.TextLineDataset`), one needs to ensure that the same file paths can be accessed on every node in the cluster.

File: docs/docs/Orca/data-parallel-processing.md [hunk 9]
```diff
@@ -3,20 +3,104 @@
...
+TensorFlow:
+```python
+def train_data_creator(config):
+    (train_feature, train_label), _ = tf.keras.datasets.mnist.load_data()
+
+    dataset = tf.data.Dataset.from_tensor_slices((train_feature, train_label))
+    dataset = dataset.repeat()
+    dataset = dataset.map(preprocess)
+    dataset = dataset.shuffle(1000)
+    dataset = dataset.batch(config["batch_size"])
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:73]: can we make it simpler?

File: docs/docs/Orca/data-parallel-processing.md [hunk 10]
```diff
@@ -3,20 +3,104 @@
...
+    return train_loader
+```
 
 ### **2. Spark Dataframes**
 Orca supports Spark Dataframes as the input to the distributed training, and as the input/output of the distributed inference. Consequently, the user can easily process large-scale dataset using Apache Spark, and directly apply AI models on the distributed (and possibly in-memory) Dataframes without data conversion or serialization. 
 
-<TODO: shown a simple example, explain the input and output dataframe format>
+```python
+rdd = sc.parallelize([(1, 2, 3), (4, 5, 6), (7, 8, 9)] * 100)
+df = rdd.toDF(["feature_1", "feature_2", "label"])
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:96]: df = spark.read.parquet("data.parquet")

File: docs/docs/Orca/data-parallel-processing.md [hunk 11]
```diff
@@ -3,20 +3,104 @@
...
+est = Estimator.from_keras(keras_model=model) # the model accept two inputs and one label
+est.fit(data=df,
+        batch_size=4,
+        epochs=1,
+        feature_cols=['user', 'item'], # specifies which column(s) to be used as inputs
+        labels_cols=['label']) # specifies which column(s) to be used as labels
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:102]:
```python
est.fit(data=df,
        feature_cols=['user', 'item'], # specifies which column(s) to be used as inputs
        labels_cols=['label']) # specifies which column(s) to be used as labels
```

File: docs/docs/Orca/data-parallel-processing.md [hunk 12]
```diff
@@ -3,20 +3,104 @@
...
+train_loader = torch.utils.data.DataLoader(
+        datasets.MNIST("/tmp/mnist", train=True, download=True,
+                       transform=transforms.Compose([
+                           transforms.ToTensor(),
+                           transforms.Normalize((0.1307,), (0.3081,))
+                       ])),
+        batch_size=batch_size, shuffle=True)
+
+est = Estimator.from_torch(model=torch_model, optimizer=torch_optim, loss=torch_criterion)
+zoo_estimator.fit(data=train_loader, epochs=1)
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:42]: est.fit(data=train_loader)

File: docs/docs/Orca/data-parallel-processing.md [hunk 13]
```diff
@@ -3,20 +3,95 @@
...
+
 
 Under the hood, Orca will automatically replicate the _TensorFlow Dataset_ or _PyTorch DataLoader_ pipeline on each node in the cluster, shard the input data, and execute the data pipelines using Apache Spark and/or Ray distributedly. 
 
-**Note:** Known limitations include: <TODO: describe known limitations>
+**Note:** Known limitations include:
+1. TensorFlow Dataset pipeline that contains transformations defined in native python function, such as `tf.py_func`, `tf.py_function`
+and `tf.numpy_function` are currently not supported.
+2. TensorFlow Dataset pipeline created from generators, such as `Dataset.from_generators` are currently not supported.
+3. For TensorFlow Dataset and Pytorch DataLoader pipelines that read from files (including tf.data.TFRecordDataset and tf.data.TextLineDataset), one needs to ensure that the same file paths can be accessed on every node in the cluster.
```
Comment (jason-dai) [inline docs/docs/Orca/data-parallel-processing.md:51]: including `tf.data.TFRecordDataset` and `tf.data.TextLineDataset`

**Comment by jason-dai** [comment]
LGTM
